### PR TITLE
Add hasExclusion API to BuildTestVerifier

### LINKED ISSUE #2172: Add hasExclusion API to BuildTestVerifier
In addition to testing for dependencies we need to test for exclusions to specific dependencies, and in specific scopes.

### PR DESCRIPTION
**Pull request by wetted**
closes #2172

### CODE REVIEW

**Comment by timyates** [comment]
I *think* (90% sure) we can support Gradle with regular expressions like this:

```diff
diff --git a/starter-core/src/test/groovy/io/micronaut/starter/build/BuildTestUtil.groovy b/starter-core/src/test/groovy/io/micronaut/starter/build/BuildTestUtil.groovy
index 9b52c8a8ad..1778fcdac0 100644
--- a/starter-core/src/test/groovy/io/micronaut/starter/build/BuildTestUtil.groovy
+++ b/starter-core/src/test/groovy/io/micronaut/starter/build/BuildTestUtil.groovy
@@ -14,7 +14,7 @@ class BuildTestUtil {
                                       Language language,
                                       TestFramework testFramework,
                                       String template) {
-        buildTool.isGradle() ? new GradleBuildTestVerifier(template, language, testFramework) : new MavenBuildTestVerifier(template, language)
+        buildTool.isGradle() ? new GradleBuildTestVerifier(buildTool, template, language, testFramework) : new MavenBuildTestVerifier(template, language)
     }
 
     static BuildTestVerifier verifier(BuildTool buildTool,
diff --git a/starter-core/src/test/groovy/io/micronaut/starter/build/gradle/GradleBuildTestVerifier.groovy b/starter-core/src/test/groovy/io/micronaut/starter/build/gradle/GradleBuildTestVerifier.groovy
index fcfe1b9555..428497a162 100644
--- a/starter-core/src/test/groovy/io/micronaut/starter/build/gradle/GradleBuildTestVerifier.groovy
+++ b/starter-core/src/test/groovy/io/micronaut/starter/build/gradle/GradleBuildTestVerifier.groovy
@@ -1,20 +1,23 @@
 package io.micronaut.starter.build.gradle
 
-
 import io.micronaut.context.exceptions.ConfigurationException
 import io.micronaut.starter.build.BuildTestVerifier
 import io.micronaut.starter.build.dependencies.Scope
+import io.micronaut.starter.options.BuildTool
 import io.micronaut.starter.options.Language
 import io.micronaut.starter.options.TestFramework
 
 import java.util.regex.Pattern
 
 class GradleBuildTestVerifier implements BuildTestVerifier {
+
+    final BuildTool buildTool
     final String template
     final Language language
     final TestFramework testFramework
 
-    GradleBuildTestVerifier(String template, Language language, TestFramework testFramework) {
+    GradleBuildTestVerifier(BuildTool buildTool, String template, Language language, TestFramework testFramework) {
+        this.buildTool = buildTool
         this.template = template
         this.language = language
         this.testFramework = testFramework
@@ -89,12 +92,27 @@ class GradleBuildTestVerifier implements BuildTestVerifier {
 
     @Override
     boolean hasExclusion(String groupId, String artifactId, String excludedGroupId, String excludedArtifactId) {
-        throw new UnsupportedOperationException("not yet implemented");
+        hasExclusion(groupId, artifactId, excludedGroupId, excludedArtifactId, Scope.COMPILE)
     }
 
     @Override
     boolean hasExclusion(String groupId, String artifactId, String excludedGroupId, String excludedArtifactId, Scope scope) {
-        throw new UnsupportedOperationException("not yet implemented");
+        Optional<String> gradleConfigurationNameOptional = GradleConfiguration.of(scope, language, testFramework, null).map { it.getConfigurationName() }
+        if (!gradleConfigurationNameOptional.isPresent()){
+            throw new ConfigurationException("cannot match " + scope + " to gradle configuration");
+        }
+        String gradleConfigurationName = gradleConfigurationNameOptional.get()
+        String pattern = /(?s).*${gradleConfigurationName}\("${groupId.replace(".", "\\.")}:${artifactId}"\)\s*\{(.+?)\}/
+        println pattern
+        def match = template =~ pattern
+        if (match.size() > 0) {
+            if (buildTool == BuildTool.GRADLE) {
+                return match[0][1].contains("exclude(group: \"${excludedGroupId}\", module: \"${excludedArtifactId}\")")
+            } else {
+                return match[0][1].contains("exclude(group = \"${excludedGroupId}\", module = \"${excludedArtifactId}\"")
+            }
+        }
+        return false
     }
 
     @Override
diff --git a/starter-core/src/test/groovy/io/micronaut/starter/build/gradle/GradleBuildTestVerifierSpec.groovy b/starter-core/src/test/groovy/io/micronaut/starter/build/gradle/GradleBuildTestVerifierSpec.groovy
index 7e7ae28c71..0531e80b11 100644
--- a/starter-core/src/test/groovy/io/micronaut/starter/build/gradle/GradleBuildTestVerifierSpec.groovy
+++ b/starter-core/src/test/groovy/io/micronaut/starter/build/gradle/GradleBuildTestVerifierSpec.groovy
@@ -1,16 +1,15 @@
 package io.micronaut.starter.build.gradle
 
+import io.micronaut.starter.options.BuildTool
 import io.micronaut.starter.options.Language
 import io.micronaut.starter.options.TestFramework
 import spock.lang.Specification
-import spock.lang.Unroll
 
 class GradleBuildTestVerifierSpec extends Specification {
 
-    @Unroll
     void "hasDependencyRegex"(boolean expected, String template, String groupId, String artifactId, String scope) {
         given:
-        GradleBuildTestVerifier verifier = new GradleBuildTestVerifier(template, Language.JAVA, TestFramework.JUNIT)
+        GradleBuildTestVerifier verifier = new GradleBuildTestVerifier(BuildTool.GRADLE, template, Language.JAVA, TestFramework.JUNIT)
 
         expect:
         expected == verifier.hasDependency(groupId, artifactId, scope)
@@ -22,6 +21,5 @@ class GradleBuildTestVerifierSpec extends Specification {
         true     | 'implementation("io.micronaut.oraclecloud:micronaut-oraclecloud-function-http:2.0.0")' |  "io.micronaut.oraclecloud" | "micronaut-oraclecloud-function-http" | GradleConfiguration.IMPLEMENTATION.toString()
         false    | 'implementation("io.micronaut.oraclecloud:micronaut-oraclecloud-function-http")'       |  "io.micronaut"             | "micronaut-oraclecloud-function-http" | GradleConfiguration.IMPLEMENTATION.toString()
         false    | 'implementation("io.micronaut.oraclecloud:micronaut-oraclecloud-function-http")'       |  "io.micronaut.oraclecloud" | "micronaut-oraclecloud-function"      | GradleConfiguration.IMPLEMENTATION.toString()
-
     }
 }
diff --git a/starter-core/src/test/groovy/io/micronaut/starter/feature/aws/DynamoDbSpec.groovy b/starter-core/src/test/groovy/io/micronaut/starter/feature/aws/DynamoDbSpec.groovy
index ac98701451..ff974006ba 100644
--- a/starter-core/src/test/groovy/io/micronaut/starter/feature/aws/DynamoDbSpec.groovy
+++ b/starter-core/src/test/groovy/io/micronaut/starter/feature/aws/DynamoDbSpec.groovy
@@ -56,10 +56,8 @@ class DynamoDbSpec extends ApplicationContextSpec implements CommandOutputFixtur
 
         then:
         verifier.hasDependency("io.micronaut.aws", "micronaut-aws-sdk-v2", Scope.COMPILE)
-        template.contains("""    implementation("software.amazon.awssdk:dynamodb") {
-                            |      exclude(group$mapNotation "software.amazon.awssdk", module$mapNotation "apache-client")
-                            |      exclude(group$mapNotation "software.amazon.awssdk", module$mapNotation "netty-nio-client")
-                            |    }""".stripMargin())
+        verifier.hasExclusion("software.amazon.awssdk", "dynamodb", "software.amazon.awssdk", "apache-client")
+        verifier.hasExclusion("software.amazon.awssdk", "dynamodb", "software.amazon.awssdk", "netty-nio-client")
         verifier.hasDependency("software.amazon.awssdk", "url-connection-client", Scope.COMPILE)
 
         where:
```

**Comment by wetted** [comment]
Thanks @timyates Nice work on the gradle part.

**Comment by timyates** [comment]
I *think* it's right 🤔 

At least I couldn't see how it was wrong in my tests

**Comment by wetted** [comment]
> I _think_ it's right 🤔
> 
> At least I couldn't see how it was wrong in my tests

It works for the one test I found. I think this seems good enough for now, and we can address corner cases if and as they come up. Thanks for helping out.

**Review by timyates — APPROVED** [review]

File: starter-core/src/test/groovy/io/micronaut/starter/build/gradle/GradleBuildTestVerifier.groovy [hunk 1]
```diff
@@ -86,6 +90,31 @@ class GradleBuildTestVerifier implements BuildTestVerifier {
...
+        Optional<String> gradleConfigurationNameOptional = GradleConfiguration.of(scope, language, testFramework, null).map { it.getConfigurationName() }
+        if (!gradleConfigurationNameOptional.isPresent()){
+            throw new ConfigurationException("cannot match " + scope + " to gradle configuration");
+        }
+        String gradleConfigurationName = gradleConfigurationNameOptional.get()
+        String pattern = /(?s).*${gradleConfigurationName}\("${groupId.replace(".", "\\.")}:${artifactId}"\)\s*\{(.+?)\}/
+        println pattern
+        def match = template =~ pattern
+        if (match.size() > 0) {
+            if (buildTool == BuildTool.GRADLE) {
```
Comment (sdelamo) [inline starter-core/src/test/groovy/io/micronaut/starter/build/gradle/GradleBuildTestVerifier.groovy]: Rewrite: 

```groovy
if (buildTool == BuildTool.GRADLE) {
    return match[0][1].contains("exclude(group: \"${excludedGroupId}\", module: \"${excludedArtifactId}\")")
 } else {
     return match[0][1].contains("exclude(group = \"${excludedGroupId}\", module = \"${excludedArtifactId}\"")
}
```

as: 

```groovy
return match[0][1].contains(buildTool == BuildTool.GRADLE 
  ? "exclude(group: \"${excludedGroupId}\", module: \"${excludedArtifactId}\")"
  : "exclude(group = \"${excludedGroupId}\", module = \"${excludedArtifactId}\"")
```

Comment (wetted) [reply starter-core/src/test/groovy/io/micronaut/starter/build/gradle/GradleBuildTestVerifier.groovy]: ping @sdelamo 

I did it a bit differently. See if that's OK
